sync/src/main/java/tech/pegasys/teku/sync/multipeer/SyncController.java

```diff
@@ -0,0 +1,143 @@
+  private final ChainSelector finalizedTargetChainSelector;
+  private final Sync finalizedSync;
+
+  /**
+   * The current sync. When empty, no sync has started, otherwise contains the details of the last
+   * started sync, which may have completed.
+   *
+   * <p>Note that this field may be accessed from off the event thread so that the sync status can
+   * be queried from any thread. It should only be written from the event thread.
+   */
+  private Optional<InProgressSync> currentSync = Optional.empty();
+
+  public SyncController(
+      final EventThread eventThread,
+      final RecentChainData recentChainData,
+      final ChainSelector finalizedTargetChainSelector,
+      final Sync finalizedSync) {
+    this.eventThread = eventThread;
+    this.recentChainData = recentChainData;
+    this.finalizedTargetChainSelector = finalizedTargetChainSelector;
+    this.finalizedSync = finalizedSync;
+  }
+
+  /**
+   * Notify Must be called on the sync event thread.
+   *
+   * @param finalizedChains the currently known finalized chains to consider
+   */
+  public void onTargetChainsUpdated(final TargetChains finalizedChains) {
+    eventThread.checkOnEventThread();
+    final Optional<InProgressSync> newFinalizedSync =
+        finalizedTargetChainSelector
+            .selectTargetChain(finalizedChains)
+            .map(this::startFinalizedSync);
+    if (newFinalizedSync.isEmpty() && isSyncActive()) {
+      return;
+    }
+    if (!isSyncActive() && newFinalizedSync.isPresent()) {
+      subscribers.deliver(SyncSubscriber::onSyncingChange, true);
+    }
+    currentSync = newFinalizedSync;
+  }
+
+  private void onSyncComplete() {
+    eventThread.checkOnEventThread();
+    if (isSyncActive()) {
+      // A different sync is now running so ignore this change.
+      return;
+    }
+    subscribers.deliver(SyncSubscriber::onSyncingChange, false);
+  }
+
+  public boolean isSyncActive() {
+    return currentSync.map(InProgressSync::isActive).orElse(false);
+  }
+
+  public SyncingStatus getSyncStatus() {
+    return currentSync.map(InProgressSync::asSyncingStatus).orElseGet(this::notSyncingStatus);
+  }
+
+  private SyncingStatus notSyncingStatus() {
+    return new SyncingStatus(false, recentChainData.getHeadSlot());
+  }
+
+  public long subscribeToSyncChanges(final SyncSubscriber subscriber) {
+    return subscribers.subscribe(subscriber);
+  }
+
+  public void unsubscribeFromSyncChanges(final long subscriberId) {
+    subscribers.unsubscribe(subscriberId);
+  }
+
+  private InProgressSync startFinalizedSync(final TargetChain chain) {
+    eventThread.checkOnEventThread();
+    if (currentSync.map(current -> current.hasSameTarget(chain)).orElse(false)) {
+      return currentSync.get();
+    }
+    final UInt64 startSlot = recentChainData.getHeadSlot();
+    final SafeFuture<Void> syncResult = finalizedSync.syncToChain(chain);
+    syncResult.thenRunAsync(this::onSyncComplete, eventThread).reportExceptions();
```

Comment:
If the target chain changes the previous sync `Future` should be exceptionally completed? If yes then we should handle this exception gracefully